Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedHandDataProvider.cs

```diff
@@ -255,41 +249,64 @@ public void Update()
             {
                 rotationDeltaEulerAngles.z = -rotationDelta;
             }
-            stateLeft.SimulateInput(mouseDelta, profile.HandJitterAmount, rotationDeltaEulerAngles);
-            stateRight.SimulateInput(mouseDelta, profile.HandJitterAmount, rotationDeltaEulerAngles);
+
+            SimulateHandInput(ref lastSimulatedTimestampLeft, HandStateLeft, isSimulatingLeft, IsAlwaysVisibleLeft, mouseDelta, rotationDeltaEulerAngles);
+            SimulateHandInput(ref lastSimulatedTimestampRight, HandStateRight, isSimulatingRight, IsAlwaysVisibleRight, mouseDelta, rotationDeltaEulerAngles);
 
             float gestureAnimDelta = profile.HandGestureAnimationSpeed * Time.deltaTime;
-            AnimateGesture(stateLeft, gestureAnimDelta);
-            AnimateGesture(stateRight, gestureAnimDelta);
+            HandStateLeft.GestureBlending += gestureAnimDelta;
+            HandStateRight.GestureBlending += gestureAnimDelta;
 
             lastMousePosition = UnityEngine.Input.mousePosition;
-
-            ApplyHandData();
         }
 
-        private void ApplyHandData()
+        /// Apply changes to one hand and update tracking
+        private void SimulateHandInput(
+            ref long lastSimulatedTimestamp,
+            SimulatedHandState state,
+            bool isSimulating,
+            bool isAlwaysVisible,
+            Vector3 mouseDelta,
+            Vector3 rotationDeltaEulerAngles)
         {
-            bool handDataChanged = false;
-            handDataChanged |= UpdateHandDataFromState(CurrentFrameLeft, stateLeft);
-            handDataChanged |= UpdateHandDataFromState(CurrentFrameRight, stateRight);
+            if (!state.IsTracked && isSimulating)
+            {
+                // Start at current mouse position
+                Vector3 mousePos = UnityEngine.Input.mousePosition;
+                state.ScreenPosition = new Vector3(mousePos.x, mousePos.y, profile.DefaultHandDistance);
+            }
 
-            if (handDataChanged)
+            if (isSimulating)
             {
-                OnHandDataChanged();
+                state.SimulateInput(mouseDelta, profile.HandJitterAmount, rotationDeltaEulerAngles);
+
+                if (isAlwaysVisible)
+                {
+                    // Toggle gestures on/off
+                    state.Gesture = ToggleGesture(state.Gesture);
+                }
+                else
+                {
+                    // Enable gesture while mouse button is pressed
+                    state.Gesture = SelectGesture();
+                }
             }
-        }
 
-        private void AnimateGesture(SimulatedHandState state, float gestureAnimDelta)
-        {
-            if (state.IsAlwaysTracked)
+            // Update tracked state of a hand.
+            // If hideTimeout value is null, hands will stay visible after tracking stops.
+            DateTime currentTime = DateTime.UtcNow;
```

Comment:
>DateTime currentTime = DateTime.UtcNow [](start = 12, length = 38)

Apparently this can be quite unprecise (~50ms intervals): [https://stackoverflow.com/questions/2143140/c-sharp-datetime-now-precision](https://stackoverflow.com/questions/2143140/c-sharp-datetime-now-precision). I'd say that for timestamps in general it's better to use Stopwatch.GetTimestamp: [https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.stopwatch.gettimestamp?view=netframework-4.7.2](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.stopwatch.gettimestamp?view=netframework-4.7.2)

Reply:
Legacy code, but good point anyway. Right now the timestamp is only used to detect if data has been changed in the same frame. I believe it's supposed to prevent multiple hand data updates in the same frame.

I would leave it as is for now, since it works alright. Will have to revisit timestamps anyway when looking into support for simulated update frequency, i.e. simulate 45fps hand updates of the Hololens instead of 60fps Unity updates. I'll add a comment to change this in future.